src/wp-includes/functions.php

```diff
@@ -8539,3 +8539,93 @@ function is_php_version_compatible( $required ) {
 function wp_fuzzy_number_match( $expected, $actual, $precision = 1 ) {
 	return abs( (float) $expected - (float) $actual ) <= $precision;
 }
+
+/**
+ * Get supported data types for meta and option values.
+ *
+ * This is a subset of the PHP data types that can be used for
+ * meta and option values. See the documentation for PHP's 'settype()'.
+ *
+ * @since 6.3.0
+ *
+ * @return array List of supported value types.
+ */
+function wp_get_db_value_types() {
+	return array(
+		'boolean',
+		'bool',
+		'integer',
+		'int',
+		'float',
+		'string',
+		'array',
+		'object',
+	);
+}
+
+/**
+ * Set the type of a meta or option value that was retrieved from the database.
+ *
+ * If `$value_type` is empty, the value is returned unchanged. If `$value_type` is not one
+ * of the expected types, a user error is triggered and the value is returned unchanged.
+ *
+ * @since 6.3.0
+ *
+ * @param mixed  $value      Raw value from the database.
+ * @param string $value_type The expected value type.
+ * @return mixed The value with the requested type.
+ */
+function wp_settype_to_value_from_db( $value, $value_type ) {
+	if ( ! empty( $value_type ) ) {
+		$value_types = wp_get_db_value_types();
+
+		if ( in_array( $value_type, $value_types, true ) ) {
+			settype( $value, $value_type );
+		} else {
+			// Improper use.
+			/* translators: The function name that returns the supported types. */
+			$message = sprintf( __( 'Only types returned by %s are supported.' ), '<code>wp_get_db_value_types()</code>' );
+			_doing_it_wrong( 'wp_settype_to_value_from_db', $message );
```

Comment:
```suggestion
			
			_doing_it_wrong(
				'wp_settype_to_value_from_db',
				sprintf(
					/* translators: %s: The function name that returns the supported types. */
					__( 'Only types returned by %s are supported.' ),
					'<code>wp_get_db_value_types()</code>'
				)
			);
```
1. We don't need to store the message in a variable.
2. By passing the `sprintf()` directly to `_doing_it_wrong()`, we can have the `// Improper use.` inline comment separated from the `/* translators */` comment.
3. Added `%s:` to the `/*translators */` comment.